### ISSUE TITLE
JSON Pointer parsing

### ISSUE THREAD

**Issue by char0n** [APIDevTools/json-schema-ref-parser#296]
Hi,

I've noticed that implementation of [JSON Pointer parsing](https://github.com/APIDevTools/json-schema-ref-parser/blob/main/lib/pointer.js#L182) is not compliant with [RFC 6901](https://www.rfc-editor.org/rfc/rfc6901), nor it's safe (throws error). The issue is with [decodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent), which [trows error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Malformed_URI) in certain character sequences which are valid JSON Pointers.

**Steps to Reproduce**

```js
const Pointer = require('@apidevtools/json-schema-ref-parser/lib/pointer.js');

Pointer.parse('#/c%d');
```

---

`/c%d` - is valid JSON Pointer according to the [RFC 6901 (section 5)](https://www.rfc-editor.org/rfc/rfc6901#section-5), which has a nice example suitable for creating a test suite for the implementation:

![image](https://user-images.githubusercontent.com/193286/212658583-188ce3b6-bd32-4014-a946-8888b392f727.png)

**Remediation**

Remediation might be introduced by safe variant of `decodeURIComponent` that will look as follows:

```js
const safeDecodeURIComponent = (encodedURIComponent: string): string => {
  try {
    return decodeURIComponent(encodedURIComponent);
  } catch {
    return encodedURIComponent;
  }
};
```

This safe version at least gives us chance to evaluate the JSON Pointer against the structure, even though not all character sequences might get decoded.

**Comment by char0n** [comment]
It's also worth considering if `encodeURIComponent` and `decodeURIComponent` should be used at all as the specification says the following:

>  Evaluation of each reference token begins by decoding any escaped
   character sequence.  This is performed by first transforming any
   occurrence of the sequence `'~1'` to `'/'`, and then transforming any
   occurrence of the sequence `'~0'` to `'~'`. 

This is little ambiguous as it's first tells us to that it `begins by decoding any escaped character sequence` but then it enumerates exactly two decoding operations that needs to be performed.

**Comment by jonluca** [comment]
Agreed, I think we do a bit too much string encoding in general. Would be better to fallback to explicit implementations of these schemas. For now I've implemented it in v11.2.4

**Comment by char0n** [comment]
Thanks a lot!
